core/string/ustring.cpp

```diff
@@ -337,28 +339,27 @@ void String::append_utf32(const Span<char32_t> &p_cstr) {
 		if ((chr & 0xfffff800) == 0xd800) {
 			print_unicode_error(vformat("Unpaired surrogate (%x)", (uint32_t)chr), true);
 			*dst = _replacement_char;
+			error = ERR_PARSE_ERROR;
 			continue;
 		}
 		if (chr > 0x10ffff) {
 			print_unicode_error(vformat("Invalid unicode codepoint (%x)", (uint32_t)chr), true);
 			*dst = _replacement_char;
+			error = ERR_PARSE_ERROR;
 			continue;
 		}
 		*dst = chr;
 	}
 	*dst = 0;
+	return error;
 }
 
-// assumes the following have already been validated:
-// p_char != nullptr
-// p_length > 0
-// p_length <= p_char strlen
-// p_char is a valid UTF32 string
-void String::copy_from_unchecked(const char32_t *p_char, const int p_length) {
-	resize(p_length + 1); // + 1 for \0
-	char32_t *dst = ptrw();
-	memcpy(dst, p_char, p_length * sizeof(char32_t));
-	*(dst + p_length) = _null;
+void String::append_utf32_unchecked(const Span<char32_t> &p_span) {
+	const int prev_length = length();
```

Comment:
funny that you cche `length()` that is called twice, but not `p_span.size()` that is called 3 times.

Reply:
`p_span.size()` just compiles to an inlined field access (`p_span._ptr`). There's no need to cache it :)